File: analyzer.py
"""Core analysis module - runs 4 parallel Claude CLI calls for each dimension."""

import json
import logging
import os
import subprocess
import tempfile
from concurrent.futures import ThreadPoolExecutor, as_completed
from pathlib import Path
from typing import Dict, List, Optional

logger = logging.getLogger(__name__)

DIMENSIONS = ["scalability", "stability", "maintainability", "security"]
PROMPTS_DIR = Path(__file__).parent / "prompts"
CLAUDE_CLI = "/Users/grooverider/.local/bin/claude"


def _build_env() -> dict:
    """Build clean environment for Claude CLI subprocess (remove nested session vars)."""
    env = os.environ.copy()
    env.pop("CLAUDECODE", None)
    env.pop("CLAUDE_CODE_ENTRYPOINT", None)
    return env


def _load_prompt(dimension: str) -> str:
    """Load prompt template for a given dimension."""
    prompt_file = PROMPTS_DIR / f"{dimension}.txt"
    logger.debug("Loading prompt for dimension=%s from %s", dimension, prompt_file)
    return prompt_file.read_text(encoding="utf-8")


def _extract_json(text: str) -> Optional[dict]:
    """Extract JSON from Claude CLI output, handling markdown code blocks."""
    text = text.strip()

    # Try direct parse first
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    # Try extracting from markdown code block
    if "```" in text:
        lines = text.split("\n")
        json_lines = []
        in_block = False
        for line in lines:
            if line.strip().startswith("```"):
                if in_block:
                    break
                in_block = True
                continue
            if in_block:
                json_lines.append(line)
        if json_lines:
            try:
                return json.loads("\n".join(json_lines))
            except json.JSONDecodeError:
                pass

    # Try finding JSON object boundaries
    start = text.find("{")
    end = text.rfind("}")
    if start != -1 and end != -1 and end > start:
        try:
            return json.loads(text[start:end + 1])
        except json.JSONDecodeError:
            pass

    return None


TIMEOUT_SECONDS = 600  # 10 minutes per dimension


def _run_claude_cli(dimension: str, repo_path: str, prompt: str) -> dict:
    """Execute a single Claude CLI call for a dimension.

    Returns:
        Parsed JSON result dict, or error dict on failure.
    """
    try:
        result = subprocess.run(
            [CLAUDE_CLI, "--print", "--model", "sonnet", "-p", prompt],
            cwd=repo_path,
            env=_build_env(),
            stdin=subprocess.DEVNULL,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=TIMEOUT_SECONDS,
            text=True,
        )
        logger.debug(
            "Claude CLI finished: dimension=%s, exit_code=%d, stdout_len=%d",
            dimension, result.returncode, len(result.stdout),
        )

        if result.returncode != 0:
            logger.error(
                "Claude CLI error: dimension=%s, stderr=%s",
                dimension, result.stderr[:500],
            )
            return {
                "dimension": dimension,
                "error": f"Claude CLI exited with code {result.returncode}",
                "stderr": result.stderr[:500],
            }

        parsed = _extract_json(result.stdout)
        if parsed is None:
            logger.error(
                "Failed to parse JSON: dimension=%s, output=%s",
                dimension, result.stdout[:500],
            )
            return {
                "dimension": dimension,
                "error": "Failed to parse JSON from Claude output",
                "raw_output": result.stdout[:1000],
            }

        logger.info(
            "Analysis complete: dimension=%s, score=%.1f",
            dimension, parsed.get("overall_score", 0),
        )
        return parsed

    except subprocess.TimeoutExpired:
        logger.error("Claude CLI timeout: dimension=%s (limit=%ds)", dimension, TIMEOUT_SECONDS)
        return {
            "dimension": dimension,
            "error": f"Analysis timed out ({TIMEOUT_SECONDS}s limit)",
        }
    except Exception as e:
        logger.error("Unexpected error: dimension=%s, error=%s", dimension, e)
        return {
            "dimension": dimension,
            "error": str(e),
        }


def analyze_dimension(dimension: str, repo_path: str) -> dict:
    """Run Claude CLI analysis for a single dimension with 1 retry on failure.

    Args:
        dimension: One of 'scalability', 'stability', 'maintainability', 'security'
        repo_path: Absolute path to the cloned repository

    Returns:
        Parsed JSON result dict, or error dict on failure
    """
    logger.info("Starting analysis: dimension=%s, repo=%s", dimension, repo_path)
    prompt = _load_prompt(dimension)

    result = _run_claude_cli(dimension, repo_path, prompt)

    # Retry once if first attempt failed
    if "error" in result:
        logger.warning(
            "First attempt failed for dimension=%s, retrying... (error: %s)",
            dimension, result.get("error", "unknown"),
        )
        result = _run_claude_cli(dimension, repo_path, prompt)
        if "error" not in result:
            logger.info("Retry succeeded for dimension=%s", dimension)
        else:
            logger.error(
                "Retry also failed for dimension=%s: %s",
                dimension, result.get("error", "unknown"),
            )

    return result


def analyze_all(repo_path: str) -> List[dict]:
    """Run all 4 dimension analyses in parallel.

    Args:
        repo_path: Absolute path to the cloned repository

    Returns:
        List of 4 result dicts (one per dimension)
    """
    logger.info("Starting parallel analysis for all dimensions: repo=%s", repo_path)
    results = []

    with ThreadPoolExecutor(max_workers=4) as executor:
        future_to_dim = {
            executor.submit(analyze_dimension, dim, repo_path): dim
            for dim in DIMENSIONS
        }
        for future in as_completed(future_to_dim):
            dim = future_to_dim[future]
            try:
                result = future.result()
                results.append(result)
                logger.info("Collected result: dimension=%s", dim)
            except Exception as e:
                logger.error("Future failed: dimension=%s, error=%s", dim, e)
                results.append({"dimension": dim, "error": str(e)})

    logger.info("All analyses complete: %d results collected", len(results))
    return results

File: src/lib/analyzer.ts
import { execFile, execSync } from "child_process";
import fs from "fs";
import path from "path";
import { completeJob, updateJobStatus } from "./db";
import { getGrade } from "./grades";
import { logger } from "@/utils/logger";

const DIMENSIONS = [
  "scalability",
  "stability",
  "maintainability",
  "security",
] as const;

type Dimension = (typeof DIMENSIONS)[number];

const PROMPTS_DIR = path.resolve(process.cwd(), "prompts");
const TIMEOUT_MS = 600_000; // 10 minutes

function findClaudeCli(): string {
  if (process.env.CLAUDE_CLI_PATH) {
    return process.env.CLAUDE_CLI_PATH;
  }
  try {
    const resolved = execSync("which claude", { encoding: "utf-8" }).trim();
    if (resolved) {
      logger.info(`[analyzer] Claude CLI found at: ${resolved}`);
      return resolved;
    }
  } catch {
    // which failed
  }
  const fallbacks = [
    path.join(process.env.HOME || "", ".local", "bin", "claude"),
    "/usr/local/bin/claude",
    "/usr/bin/claude",
  ];
  for (const p of fallbacks) {
    if (fs.existsSync(p)) {
      logger.info(`[analyzer] Claude CLI found at fallback: ${p}`);
      return p;
    }
  }
  logger.error("[analyzer] Claude CLI not found anywhere");
  return "claude"; // last resort: hope it's in PATH
}

let _claudeCli: string | null = null;
function getClaudeCli(): string {
  if (!_claudeCli) {
    _claudeCli = findClaudeCli();
  }
  return _claudeCli;
}

function loadPrompt(dimension: Dimension): string {
  const promptPath = path.join(PROMPTS_DIR, `${dimension}.txt`);
  return fs.readFileSync(promptPath, "utf-8");
}

function extractJson(text: string): Record<string, unknown> | null {
  const trimmed = text.trim();

  // Try direct parse
  try {
    return JSON.parse(trimmed);
  } catch {
    // continue
  }

  // Try extracting from markdown code block
  if (trimmed.includes("```")) {
    const lines = trimmed.split("\n");
    const jsonLines: string[] = [];
    let inBlock = false;
    for (const line of lines) {
      if (line.trim().startsWith("```")) {
        if (inBlock) break;
        inBlock = true;
        continue;
      }
      if (inBlock) jsonLines.push(line);
    }
    if (jsonLines.length > 0) {
      try {
        return JSON.parse(jsonLines.join("\n"));
      } catch {
        // continue
      }
    }
  }

  // Try finding JSON object boundaries
  const start = trimmed.indexOf("{");
  const end = trimmed.lastIndexOf("}");
  if (start !== -1 && end !== -1 && end > start) {
    try {
      return JSON.parse(trimmed.slice(start, end + 1));
    } catch {
      // continue
    }
  }

  return null;
}

function runClaudeCli(
  dimension: Dimension,
  repoPath: string,
  prompt: string
): Promise<Record<string, unknown>> {
  return new Promise((resolve) => {
    // Build clean env (remove nested Claude session vars)
    const env = { ...process.env };
    delete env.CLAUDECODE;
    delete env.CLAUDE_CODE_ENTRYPOINT;

    const child = execFile(
      getClaudeCli(),
      ["--print", "--model", "sonnet", "-p", prompt],
      {
        cwd: repoPath,
        env,
        timeout: TIMEOUT_MS,
        maxBuffer: 10 * 1024 * 1024,
      },
      (error, stdout, stderr) => {
        if (error) {
          if (error.killed || error.code === "ETIMEDOUT") {
            logger.error(
              `[analyzer] Claude CLI timeout: dimension=${dimension}`
            );
            resolve({
              dimension,
              error: `Analysis timed out (${TIMEOUT_MS / 1000}s limit)`,
            });
            return;
          }
          logger.error(
            `[analyzer] Claude CLI error: dimension=${dimension}, error=${error.message}, stderr=${stderr?.slice(0, 500)}`
          );
          resolve({
            dimension,
            error: `Claude CLI failed: ${error.message}`,
          });
          return;
        }

        const parsed = extractJson(stdout);
        if (!parsed) {
          logger.error(
            `[analyzer] Failed to parse JSON: dimension=${dimension}, output=${stdout.slice(0, 500)}`
          );
          resolve({
            dimension,
            error: "Failed to parse JSON from Claude output",
            raw_output: stdout.slice(0, 1000),
          });
          return;
        }

        logger.info(
          `[analyzer] Analysis complete: dimension=${dimension}, score=${(parsed as Record<string, number>).overall_score?.toFixed(1)}`
        );
        resolve(parsed);
      }
    );

    // Ensure stdin is closed
    child.stdin?.end();
  });
}

async function analyzeDimension(
  dimension: Dimension,
  repoPath: string
): Promise<Record<string, unknown>> {
  logger.info(
    `[analyzer] Starting analysis: dimension=${dimension}, repo=${repoPath}`
  );
  const prompt = loadPrompt(dimension);

  let result = await runClaudeCli(dimension, repoPath, prompt);

  // Retry once if first attempt failed
  if ("error" in result) {
    logger.warn(
      `[analyzer] First attempt failed for dimension=${dimension}, retrying...`
    );
    result = await runClaudeCli(dimension, repoPath, prompt);
    if (!("error" in result)) {
      logger.info(`[analyzer] Retry succeeded for dimension=${dimension}`);
    } else {
      logger.error(
        `[analyzer] Retry also failed for dimension=${dimension}`
      );
    }
  }

  return result;
}

async function analyzeAll(
  repoPath: string
): Promise<Record<string, unknown>[]> {
  logger.info(
    `[analyzer] Starting parallel analysis for all dimensions: repo=${repoPath}`
  );

  const results = await Promise.all(
    DIMENSIONS.map((dim) => analyzeDimension(dim, repoPath))
  );

  logger.info(
    `[analyzer] All analyses complete: ${results.length} results collected`
  );
  return results;
}

/** Run analysis in background. Returns immediately. */
export function analyzeInBackground(
  jobId: string,
  codeDirPath: string
): void {
  (async () => {
    logger.info(`[analyzer] Background analysis started: job_id=${jobId}`);
    updateJobStatus(jobId, "analyzing");

    try {
      const results = await analyzeAll(codeDirPath);

      // Calculate summary
      const valid = results.filter(
        (r) => !("error" in r) && "overall_score" in r
      );
      const failed = results.filter((r) => "error" in r);

      const scores: Record<string, number> = {};
      for (const r of valid) {
        const dim = (r.dimension as string) || "unknown";
        scores[dim] = (r.overall_score as number) || 0;
      }

      const failedDims = failed.map(
        (r) => (r.dimension as string) || "unknown"
      );

      if (failedDims.length > 0) {
        logger.warn(
          `[analyzer] Some dimensions failed: ${failedDims.join(", ")} (job_id=${jobId})`
        );
      }

      const scoreValues = Object.values(scores);
      const avgScore =
        scoreValues.length > 0
          ? scoreValues.reduce((a, b) => a + b, 0) / scoreValues.length
          : 0;

      const [grade, gradeDesc] = getGrade(avgScore);

      const weakestDim =
        scoreValues.length > 0
          ? Object.entries(scores).reduce((a, b) =>
              a[1] < b[1] ? a : b
            )[0]
          : null;

      const summary = {
        scores,
        average: Math.round(avgScore * 100) / 100,
        grade,
        grade_desc: gradeDesc,
        weakest: weakestDim,
        weakest_score: weakestDim ? scores[weakestDim] : 0,
        failed_dimensions: failedDims,
      };

      completeJob(
        jobId,
        summary,
        results as Record<string, unknown>[]
      );
      logger.info(
        `[analyzer] Analysis complete: job_id=${jobId}, grade=${grade}`
      );
    } catch (e) {
      const errorMsg = e instanceof Error ? e.message : String(e);
      logger.error(
        `[analyzer] Analysis failed: job_id=${jobId}, error=${errorMsg}`
      );
      updateJobStatus(jobId, "failed", errorMsg);
    } finally {
      // Cleanup code directory
      logger.info(`[analyzer] Cleaning up code directory: ${codeDirPath}`);
      fs.rmSync(codeDirPath, { recursive: true, force: true });
    }
  })();
}
